Test/Unit/Model/Import/CreateRatesConsumerTest.php

```diff
@@ -0,0 +1,284 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Taxjar\SalesTax\Test\Unit\Model\Import;
+
+use Magento\AsynchronousOperations\Model\Operation;
+use Magento\Config\Model\ResourceModel\Config\Data\Collection;
+use Magento\Config\Model\ResourceModel\Config\Data\CollectionFactory;
+use Magento\Framework\App\Config\ScopeConfigInterface;
+use Magento\Framework\DataObject;
+use Magento\Framework\DB\LoggerInterface;
+use Magento\Framework\EntityManager\EntityManager;
+use Magento\Framework\Exception\LocalizedException;
+use Magento\Framework\Serialize\SerializerInterface;
+use Taxjar\SalesTax\Model\Configuration as TaxjarConfig;
+use Taxjar\SalesTax\Model\Import\CreateRatesConsumer;
+use Taxjar\SalesTax\Model\Import\RateFactory;
+use Taxjar\SalesTax\Model\Import\RuleFactory;
+use Taxjar\SalesTax\Test\Unit\UnitTestCase;
+
+class CreateRatesConsumerTest extends UnitTestCase
+{
+    /**
+     * @var SerializerInterface
+     */
+    protected $serializer;
+
+    /**
+     * @var ScopeConfigInterface
+     */
+    protected $scopeConfig;
+
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+
+    /**
+     * @var EntityManager $entityManager
+     */
+    protected $entityManager;
+
+    /**
+     * @var TaxjarConfig
+     */
+    protected $taxjarConfig;
+
+    /**
+     * @var RateFactory
+     */
+    protected $rateFactory;
+
+    /**
+     * @var RuleFactory
+     */
+    protected $ruleFactory;
+
+    /**
+     * @var Operation|null
+     */
+    protected $operation;
+
+    /**
+     * @var array|null
+     */
+    protected $payload;
+
+    /**
+     * @var array|null
+     */
+    protected $rates;
+
+    /**
+     * @var CollectionFactory
+     */
+    private $configCollection;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $this->serializer = $this->createMock(SerializerInterface::class);
+        $this->scopeConfig = $this->createMock(ScopeConfigInterface::class);
+        $this->logger = $this->createMock(LoggerInterface::class);
+        $this->entityManager = $this->createMock(EntityManager::class);
+        $this->taxjarConfig = $this->createMock(TaxjarConfig::class);
+        $this->rateFactory = $this->getMockBuilder(RateFactory::class)->disableOriginalConstructor()->setMethods(['create'])->getMock();
+        $this->ruleFactory = $this->getMockBuilder(RuleFactory::class)->disableOriginalConstructor()->setMethods(['create'])->getMock();
+        $this->configCollection = $this->getMockBuilder(CollectionFactory::class)->disableOriginalConstructor()->setMethods(['create'])->getMock();
+    }
+
+    public function testSetCustomerTaxClasses()
```

Comment:
For some of the test methods I'm not sure exactly what is being tested. It appears to me like the ones whose only assertion is assertInstanceOf are only testing that the sut can call a provide method and are not actually testing any expected outcomes. I'm not sure if they are really necessary.

In general I don't think private methods should be unit tested. Maybe an approach more like what you did in some of the other test classes would be of more value. Where you setup the mocks, call the public method and use the mock dependencies of the SUT to verify that the indented end result was achieved. Perhaps something like this:

$operationStub = $this->createMock(OperationInterface::class);
$operationStub->method('getSerializedData')->willReturn(whatever test data);
$rateMock->expects($this->once())->method('create')->with(whatever expected rate data)
$sut = $this->getTestSubject();
$sut->process($operationStub);

Reply:
Hey @dallendalton - So I've renamed the most of the tests in this file to be a little more verbose.
 
Some of the first tests of setters like this one are smokey (and I can remove), but most of the tests that are asserting the return self are testing the "happy" path, and there is another test for when an exception is thrown.

This is one of the classes that we discussed would be a pain to test because all of the methods are private or protected and it doesn't have a public interface aside from the abstract class it extends from. The full functionality is validated in the Integration test for this class, but the kind of unit test described where we mock ALL of the dependency functionalities and pass them thru the constructor would end up with a lot of test setup that is super coupled to the code.

Maybe we can discuss this one in sync when you get a chance?